libsolutil/Dominator.h

```diff
@@ -0,0 +1,377 @@
+				predecessors[m_dfsIndexByVertexId[w.id]].insert(m_dfsIndexByVertexId[_v.id]);
+			});
+		};
+		dfs(_entry, dfs);
+		solAssert(dfIdx == m_verticesInDFSOrder.size());
+		solAssert(dfIdx == visited.size());
+		solAssert(dfIdx == semi.size());
+		solAssert(dfIdx == label.size());
+
+		size_t numVertices = visited.size();
+		if (predecessors.size() < numVertices)
+			predecessors.resize(numVertices);
+		if (parent.size() < numVertices)
+			parent.resize(numVertices);
+
+		// ancestor(w): The highest ancestor of a vertex ``w`` in the dominator tree used
+		// for path compression.
+		std::vector<DfsIndex> ancestor(numVertices, std::numeric_limits<size_t>::max());
+		// bucket(w): a set of vertices whose semidominator is ``w``
+		// The index of the array represents the vertex's ``dfIdx``
+		std::vector<std::deque<DfsIndex>> bucket(numVertices);
+		// idom(w): the vertex Id of the immediate dominator of ``w``
+		std::vector<DfsIndex> idom(numVertices, std::numeric_limits<size_t>::max());
+
+		// ``link`` adds an edge to the virtual forest.
+		// It copies the ``parent`` index of the vertex ``w`` to the ancestor array to limit the search path upwards.
+		// TODO: implement sophisticated link-eval algorithm as shown in pg 132
+		// See: https://www.cs.princeton.edu/courses/archive/spr03/cs423/download/dominators.pdf
+		auto link = [&](DfsIndex _parentIdx, DfsIndex _wIdx)
+		{
+			ancestor[_wIdx] = _parentIdx;
+		};
+
+		// ``eval`` computes the path compression.
+		// Finds ancestor with lowest semi-dominator DFS number (i.e. index).
+		auto eval = [&](DfsIndex _vIdx) -> DfsIndex
+		{
+			if (ancestor[_vIdx] != std::numeric_limits<size_t>::max())
+			{
+				compressPath(ancestor, label, semi, _vIdx);
+				return label[_vIdx];
+			}
+			return _vIdx;
+		};
+
+		auto toDfsIndex = [&](VId const& _vId) { return m_dfsIndexByVertexId[_vId]; };
+		// Process the vertices in decreasing order of the DFS number
+		for (DfsIndex wIdx: m_verticesInDFSOrder | ranges::views::reverse | ranges::views::transform(toDfsIndex))
+		{
+			// step 3
+			// NOTE: this is an optimization, i.e. performing the step 3 before step 2.
+			// The goal is to process the bucket in the beginning of the loop for the vertex ``w``
+			// instead of ``parent[w]`` in the end of the loop as described in the original paper.
+			// Inverting those steps ensures that a bucket is only processed once and
+			// it does not need to be erased.
+			// The optimization proposal is available here: https://jgaa.info/index.php/jgaa/article/view/paper119/2847 pg.77
+			for (DfsIndex vIdx: bucket[wIdx])
+			{
+				DfsIndex uIdx = eval(vIdx);
+				solAssert(uIdx <= vIdx);
+				idom[vIdx] = (semi[uIdx] < semi[vIdx]) ? uIdx : wIdx;
+			}
+
+			// step 2
+			for (DfsIndex vIdx: predecessors[wIdx])
+			{
+				DfsIndex uIdx = eval(vIdx);
+				solAssert(uIdx <= vIdx);
+				if (semi[uIdx] < semi[wIdx])
+					semi[wIdx] = semi[uIdx];
+			}
+			bucket[semi[wIdx]].emplace_back(wIdx);
```

Comment:
It is possible for `wIdx` and `semi[wIdx]` to be equal. When that happens, we're putting `wIdx` in a bucket that has already been processed. Doesn't this break something?

Reply:
Ah, wait, it should not really be possible. Turns out it happens here only for the root vertex. So it's probably fine, but we should assert that:

```c++
solAssert(semi[wIdx] < wIdx || wIdx == 0);
```

Reply:
I added this and some other asserts in a fixup.